prow/test/integration/internal/fakepubsub/fakepubsub.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fakepubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/option"
+	"k8s.io/test-infra/prow/pubsub/subscriber"
+)
+
+type PubSubMessageForSub struct {
+	Attributes map[string]string
+	Data       subscriber.ProwJobEvent
+}
+
+type Client struct {
+	projectID    string
+	pubsubClient *pubsub.Client
+}
+
+func NewClient(projectID, pubsubEmulatorHost string) (*Client, error) {
+	// Force this client to always talk to the fakepubsub service running in the
+	// test KIND cluster. This is global to the test process (all integration
+	// tests). If you need your tests to talk to the real Pub/Sub for whatever
+	// reason you can unset this variable and set it back in your test. However
+	// you must be careful with parallelism (i.e., avoid calling t.Parallel()).
+	//
+	// Although we can technically set this elsewhere (perhaps in TestMain()),
+	// we choose to set it here to force the client to talk to the fakepubsub
+	// service, and to make the intent of the code here more explicit.
+	os.Setenv("PUBSUB_EMULATOR_HOST", pubsubEmulatorHost)
+
+	client, err := pubsub.NewClient(context.Background(), projectID, option.WithoutAuthentication())
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create pubsub client to project %q: %v", projectID, err)
+	}
+
+	return &Client{
+		projectID:    projectID,
+		pubsubClient: client,
+	}, nil
+}
+
+// PublishMessage creates a Pub/Sub message that sub understands (to create a
+// ProwJob). The podName parameter is used by the integration tests;
+// specifically, each test case invocation generates a UUID which is used as the
+// name of the ProwJob CR. Then when the test pod is created, it is also named
+// with the same UUID. This makes checking for the creation of jobs and pods
+// very easy in the tests.
+func (c *Client) PublishMessage(msg PubSubMessageForSub, topicID string) error {
+	bytes, err := json.Marshal(msg.Data)
+	if err != nil {
+		return fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	t := c.pubsubClient.Topic(topicID)
+	ctx := context.Background()
```

Comment:
same as last comment, might want to pass in a context

Reply:
Done